[WIP] Unboxing for C++ backend

benpye opened the pull request:
This change set adds limited unboxing support for the C++ backend. Calling interface methods on boxed value types is functional, as is calling methods such as `Convert.ToInt32` with a boxed value type. Enums (as in #2884) still fails, I need to look at what is going on with `FindInterfaceMethodImplementationTarget`.

@jkotas 

Working examples:
```c#
using System;

internal class Program
{
    interface IMyInterface
    {
        void MyMethod();
    }
    struct MyValueType : IMyInterface
    {
        void IMyInterface.MyMethod()
        {
            Console.WriteLine("IMyInterface.MyMethod");
        }
    }
    private static void Main()
    {
        object o = new MyValueType();
        ((IMyInterface)o).MyMethod();
    }
}
```

```c#
using System;

internal class Program
{
    private static void Main()
    {
        Console.WriteLine(Convert.ToInt32((object)1));
    }
}
```

dnfclas commented:
@benpye,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

MichalStrehovsky commented on src/ILCompiler.CppCodeGen/src/CppCodeGen/CppWriter.cs:
```diff
@@ -1217,6 +1239,37 @@ private void OutputExternCSignatures()
+        /// </summary>
+        /// <param name="unboxingStubNode">The unboxing node to be output</param>
+        /// <param name="methodImplementations">The buffer in which to write out the C++ code</param>
+        private void OutputUnboxNode(UnboxingStubNode unboxingStubNode)
```
Would it be possible to just allocate a normal `CppMethodCodeNode` for these, where `_methodCode` is this code and `_dependencies` points to the `CppMethodCodeNode` of the target?

It would fix the TODO here: https://github.com/dotnet/corert/blob/e391acfbb7fb440f4b5469623f33a4b5d23be369/src/ILCompiler.CppCodeGen/src/Compiler/DependencyAnalysis/CppCodegenNodeFactory.cs#L35

It feels odd to allocate the `UnboxingStubNode` (which represents a machine code helper) and then try to patch things up later.

benpye replied:
I was thinking of using a derived type, we need to differentiate at line 851, unless all uses of `GetCppMethodName` are replaced by `method.GetMangledName`

benpye replied:
The unboxing stubs define their own mangled name. We could derive `CppMethodCodeNode`, providing the unboxing names. Would have to swap `GetCppMethodName` for `method.GetMangledName` in several places I think.

benpye commented:
Will address the two comments.

Is it possible to trace through the RyuJIT compiled code with any symbols. In the struct case it seems that `FindImplSlotForCurrentType` (at https://github.com/dotnet/corert/blob/master/src/Runtime.Base/src/System/Runtime/DispatchResolve.cs#L51 ) fails until `pCur` is `NULL`, and we return 0. Would be useful to see the expected behaviour to compare.

jkotas commented on src/ILCompiler.CppCodeGen/src/CppCodeGen/CppWriter.cs:
```diff
@@ -232,6 +232,18 @@ public void AppendCppMethodCallParamList(CppGenerationBuffer sb, MethodDesc meth
+                    System.Diagnostics.Debug.Assert(hasThis);
+
+                    var thisType = method.OwningType;
+                    if (thisType.IsValueType)
```
I think that this should be assert instead as well - unboxing stubs are only valid for value types.

jkotas commented:
@benpye Do you plan to do more changes for this? I think this change is improvement as is - I would be happy to merge it. The cleanup around the unboxing stub node define can be done separately.

benpye commented:
@jkotas Happy to merge this as is, as you say, already an improvement. Have addressed your comments and squashed. Can work on cleaning up and what is going on with the enum case.

jkotas commented:
Thanks!
